[BOLT] Skip out-of-range pending relocations

paschalis-mpeis opened the pull request:
When a pending relocation is created it is also marked whether it is
optional or not. It can be optional when such relocation is added as
part of an optimization (i.e., `scanExternalRefs`).

When bolt tries to `flushPendingRelocations`, it safely skips any
optional relocations that cannot be encoded.

Background:
BOLT, as part of scanExternalRefs, identifies external references from
calls and creates some pending relocations for them. Those when flushed
will update references to point to the optimized functions. This
optimization can be disabled using `--no-scan`.

BOLT can assert if any of these pending relocations cannot be encoded.

This patch does not disable this optimization but instead selectively
applies it given that a pending relocation is optional.

maksfb commented on bolt/lib/Core/BinarySection.cpp:
```diff
@@ -165,11 +165,17 @@ void BinarySection::flushPendingRelocations(raw_pwrite_stream &OS,
     if (Reloc.Symbol)
       Value += Resolver(Reloc.Symbol);
 
+    if (!Relocation::canEncodeValue(Reloc.Type, Value,
```
Agreed, we need a predicate for skipping. I.e. define a condition when it's safe to skip processing relocations.

paschalis-mpeis replied:
`addPendingRelocation` now becomes the only way now to add a pending relocation (Parent PR #123635). If it comes from the `scanExternalRefs` optimization, then it is marked as optional. Finally, those relocations can be safely skipped when it's time to flush them.

paschalis-mpeis committed:
BOLT asserts on out-of-range pending relocs

paschalis-mpeis committed:
[BOLT] Skip flushing optional out-of-range pending relocs

When a pending relocation is created it is also marked whether it is
optional or not. It can be optional when such relocation is added as
part of an optimization (i.e., `scanExternalRefs`).

When bolt tries to `flushPendingRelocations`, it safely skips any
optional relocations that cannot be encoded.

Background:
BOLT, as part of scanExternalRefs, identifies external references from
calls and creates some pending relocations for them. Those when flushed
will update references to point to the optimized functions. This
optimization can be disabled using `--no-scan`.

BOLT can assert if any of these pending relocations cannot be encoded.

This patch does not disable this optimization but instead selectively
applies it given that a pending relocation is optional.

paschalis-mpeis commented:
Forced-pushed to add the missing code. Also, this PR on now stacked top of #123635.
Thanks for the comments @maks. I am not sure if your [concern](https://github.com/llvm/llvm-project/issues/116817#issuecomment-2602866672) on the issue still stands or not.

paschalis-mpeis committed:
Ensure PatchEntries runs

Allow skipping pending relocations only when `-force-patch` is set.
Otherwise, exit with a relevant message.

paschalis-mpeis commented:
- force-push to stack this PR on top of #127812.
- add code & test to ensure that we skip pending relocations only when `-force-patch` was set

maksfb commented on bolt/lib/Core/BinarySection.cpp:
```diff
@@ -197,6 +218,10 @@ void BinarySection::flushPendingRelocations(raw_pwrite_stream &OS,
 
   clearList(PendingRelocations);
+
+  if (SkippedPendingRelocations > 0)
+    LLVM_DEBUG(dbgs() << "BOLT-INFO: Skipped " << SkippedPendingRelocations
+                      << " pending relocations as they were out of range\n");
```
```suggestion
  if (SkippedPendingRelocations > 0 && opts::Verbosity >= 1) {
    BC.outs() << "BOLT-INFO: skipped " << SkippedPendingRelocations
                     << " out-of-range optional relocations\n";
  }
```

maksfb reviewed (commented):
Thanks for working on this! Overall, looks good to me.

Let's add `Optional` flag to the `Relocation`. We can make `Relocation::Type` `uint32_t` and then the new flag will come with zero overhead.

paschalis-mpeis committed:
Addressing reviewers

paschalis-mpeis commented:
Thanks for your review Maksim.

Forced-push to rebase on top of #133085 and add commit b4e906c to  address reviewers.

maksfb reviewed (commented):
Looks good! Minor nits.

paschalis-mpeis committed:
Addressing reviewers (2)

paschalis-mpeis commented:
Hey Maksim,

Thanks for your review. Let me know if you have any further suggestions.

maksfb commented on bolt/lib/Core/BinarySection.cpp:
```diff
@@ -174,11 +175,30 @@ void BinarySection::flushPendingRelocations(raw_pwrite_stream &OS,
+      // relocations are flushed. Otherwise, PatchEntries should run.
+      if (!opts::ForcePatch) {
+        BC.errs()
+            << "BOLT-ERROR: Cannot encode relocation for symbol "
```
nit:
```suggestion
            << "BOLT-ERROR: cannot encode relocation for symbol "
```

maksfb reviewed (approved):
Thanks, Paschalis. Looks great! Please address a tiny nit before commit.

paschalis-mpeis committed:
Addressing reviewers (3)

paschalis-mpeis commented:
Great, thanks. I'll wait a bit for other reviewers and merge by the EOD.
